File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    int first_digit = 0;
    int second_digit = 0;
    int digit_length = 0;

    long number = get_long("Number: ");
    long copy = number;

    while (copy)
    {
        digit_length += 1;
        copy /= 10;
    }

    int first_digit_index = digit_length - 1;
    int second_digit_index = digit_length - 2;

    copy = number;

    for (int i = 0; i < digit_length; i++)
    {
        if (i == first_digit_index)
        {
            first_digit = copy;
        }
        if (i == second_digit_index)
        {
            second_digit = copy % 10;
        }
        copy /= 10;
    }

    int total = 0;
    int digit_num_from_right = 1;
    while (number)
    {
        int current_digit = number % 10;

        if (digit_num_from_right % 2 == 0)
        {
            int current_doubled = current_digit * 2;
            while (current_doubled)
            {
                total += current_doubled % 10;
                current_doubled /= 10;
            }
        }
        else
        {
            total += current_digit;
        }
        digit_num_from_right += 1;
        number /= 10;
    }

    if (total % 10 == 0)
    {
        if ((digit_length == 13 || digit_length == 16) && first_digit == 4)
        {
            printf("VISA\n");
        }

        else if (digit_length == 16 && first_digit == 5 && second_digit > 0 && second_digit < 6)
        {
            printf("MASTERCARD\n");
        }

        else if (digit_length == 15 && first_digit == 3 && (second_digit == 4 || second_digit == 7))
        {
            printf("AMEX\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }

    else
    {
        printf("INVALID\n");
    }
}

File: pset6/credit/credit.py
from cs50 import get_string


def main():
    number = get_string("Number: ")

    isAmex = checkAmex(number)
    isMasterCard = checkMasterCard(number)
    isVisa = checkVisa(number)

    if(not(isVisa or isMasterCard or isAmex)):
        print("INVALID")
        return

    isValid = luhns(number)

    if isValid:
        if(isAmex):
            print("AMEX")
            return
        if(isMasterCard):
            print("MASTERCARD")
            return
        if(isVisa):
            print("VISA")
            return
    else:
        print("INVALID")


def luhns(number):
    total = 0
    counter = 1
    for digitS in reversed(number):
        digit = int(digitS)
        if counter % 2 == 0:
            product = digit * 2
            productS = str(product)

            if len(productS) > 1:
                total += int(productS[0])
                total += int(productS[1])
            else:
                total += product
        else:
            total += digit
        counter += 1
    totalS = str(total)
    last = int(totalS[-1])
    if (last == 0):
        return True
    return False

# amex starts with 34 or 37 and is 15 digits


def checkAmex(number):
    return number[0] == '3' and (number[1] == '4' or number[1] == '7') and len(number) == 15

# mastercard starts with 51-55 and is 16 digits


def checkMasterCard(number):
    secondDigit = int(number[1])
    return number[0] == '5' and secondDigit > 0 and secondDigit < 6 and len(number) == 16

# visa starts with 4 and is 13 or 16 digits


def checkVisa(number):
    return number[0] == '4' and (len(number) == 13 or len(number) == 16)


main()

